rpcs3/Emu/Cell/lv2/sys_uart.h

```diff
@@ -1,10 +1,316 @@
+	PS3AV_RESBIT_1920x1080P                 = 0x4000,
+
+	PS3AV_MONITOR_TYPE_HDMI                 = 1,
+	PS3AV_MONITOR_TYPE_DVI                  = 2,
+
+	PS3AV_STATUS_SUCCESS                    = 0x00,
+	PS3AV_STATUS_RECEIVE_VUART_ERROR        = 0x01,
+	PS3AV_STATUS_SYSCON_COMMUNICATE_FAIL    = 0x02,
+	PS3AV_STATUS_INVALID_COMMAND            = 0x03,
+	PS3AV_STATUS_INVALID_PORT               = 0x04,
+	PS3AV_STATUS_INVALID_VID                = 0x05,
+	PS3AV_STATUS_INVALID_COLOR_SPACE        = 0x06,
+	PS3AV_STATUS_INVALID_FS                 = 0x07,
+	PS3AV_STATUS_INVALID_AUDIO_CH           = 0x08,
+	PS3AV_STATUS_UNSUPPORTED_VERSION        = 0x09,
+	PS3AV_STATUS_INVALID_SAMPLE_SIZE        = 0x0A,
+	PS3AV_STATUS_FAILURE                    = 0x0B,
+	PS3AV_STATUS_UNSUPPORTED_COMMAND        = 0x0C,
+	PS3AV_STATUS_BUFFER_OVERFLOW            = 0x0D,
+	PS3AV_STATUS_INVALID_VIDEO_PARAM        = 0x0E,
+	PS3AV_STATUS_NO_SEL                     = 0x0F,
+	PS3AV_STATUS_INVALID_AV_PARAM           = 0x10,
+	PS3AV_STATUS_INVALID_AUDIO_PARAM        = 0x11,
+	PS3AV_STATUS_UNSUPPORTED_HDMI_MODE      = 0x12,
+	PS3AV_STATUS_NO_SYNC_HEAD               = 0x13,
+};
+
+class vuart_av_thread;
+
+struct ps3av_cmd
+{
+	virtual u16 get_size(vuart_av_thread &vuart, const void *pkt_buf) = 0;
+	virtual bool execute(vuart_av_thread &vuart, const void *pkt_buf, u32 reply_max_size) = 0;
+};
+
+template<u32 Size>
+class vuart_av_ringbuf
+{
+private:
+
+	atomic_t<u32> rd_ptr = 0;
+	atomic_t<u32> wr_ptr = 0;
+	u8 buf[Size + 1];
+
+public:
+
+	u32 get_free_size()
+	{
+		const u32 rd = rd_ptr.observe();
+		const u32 wr = wr_ptr.observe();
+
+		return wr >= rd ? Size - (wr - rd) : rd - wr - 1U;
+	}
+
+	u32 get_used_size()
+	{
+		return Size - get_free_size();
+	}
+
+	u32 push(const void *data, u32 size)
+	{
+		const u32 old     = wr_ptr.observe();
+		const u32 to_push = std::min(size, get_free_size());
+		const u8 *b_data  = static_cast<const u8*>(data);
+
+		if (!to_push || !data)
+		{
+			return 0;
+		}
+
+		if (old + to_push > sizeof(buf))
+		{
+			const u32 first_write_sz = sizeof(buf) - old;
+			memcpy(&buf[old], b_data, first_write_sz);
+			memcpy(&buf[0], b_data + first_write_sz, to_push - first_write_sz);
+		}
+		else
+		{
+			memcpy(&buf[old], b_data, to_push);
+		}
+
+		wr_ptr = (old + to_push) % sizeof(buf);
+
+		return to_push;
+	}
+
+	u32 pop(void *data, u32 size)
+	{
+		const u32 old    = rd_ptr.observe();
+		const u32 to_pop = std::min(size, get_used_size());
+		u8 *b_data       = static_cast<u8*>(data);
+
+		if (!to_pop || !data)
+		{
+			return 0;
+		}
+
+		if (old + to_pop > sizeof(buf))
+		{
+			const u32 first_read_sz = sizeof(buf) - old;
+			memcpy(b_data, &buf[old], first_read_sz);
+			memcpy(b_data + first_read_sz, &buf[0], to_pop - first_read_sz);
+		}
+		else
+		{
+			memcpy(b_data, &buf[old], to_pop);
+		}
+
+		rd_ptr = (old + to_pop) % sizeof(buf);
+
+		return to_pop;
+	}
+};
+
+class vuart_av_thread
+{
+public:
+
+	u8 init = 0;
```

Comment:
It seems like this is simply used to guard against double init; why not just use a boolean here? You can even call it initialized to make the code easier to read.